File: solutions/typescript/rational-numbers/1/rational-numbers.ts
export default class Rational {
  a: number;
  b: number;

  constructor(a: number, b: number) {
    this.a = a;
    this.b = b;
  }
  /**
   * The sum of two rational numbers r1 = a1/b1 and r2 = a2/b2 is
   * r1 + r2 = a1/b1 + a2/b2 = (a1 * b2 + a2 * b1) / (b1 * b2).
   */
  add(other: Rational): Rational {
    const num = this.a * other.b + other.a * this.b;
    const den = this.b * other.b;

    return new Rational(num, den).reduce();
  }

  /**
   * The difference of two rational numbers r1 = a1/b1 and r2 = a2/b2 is
   * r1 - r2 = a1/b1 - a2/b2 = (a1 * b2 - a2 * b1) / (b1 * b2
   */
  sub(other: Rational): Rational {
    const num = this.a * other.b - other.a * this.b;
    const den = this.b * other.b;

    return new Rational(num, den).reduce();
  }

  /**
   * The product (multiplication) of two rational numbers r1 = a1/b1 and r2 = a2/b2
   * is r1 * r2 = (a1 * a2) / (b1 * b2)
   */
  mul(other: Rational): Rational {
    const num = this.a * other.a;
    const den = this.b * other.b;

    return new Rational(num, den).reduce();
  }

  /**
   * Dividing a rational number r1 = a1/b1 by another r2 = a2/b2 is
   * r1 / r2 = (a1 * b2) / (a2 * b1) if a2 * b1 is not zero.
   */
  div(other: Rational): Rational {
    let num = this.a * other.b;
    let den = other.a * this.b;

    return new Rational(num, den).reduce();
  }

  abs(): Rational {
    const num = this.a < 0 ? this.a * -1 : this.a;
    const den = this.b < 0 ? this.b * -1 : this.b;

    return new Rational(num, den).reduce();
  }

  exprational(exp: number): Rational {
    const num = this.a ** exp;
    const den = this.b ** exp;

    return new Rational(num, den).reduce();
  }

  expreal(value: number): number {
    const exp = this.a / this.b;

    return value ** exp;
  }

  reduce(): Rational {
    const div = gcd(this.a, this.b);

    let num = this.a / div;
    let den = this.b / div;

    if (den < 1) {
      num *= -1;
      den *= -1;
    }

    return new Rational(num, den);
  }
}

function gcd(a: number, b: number): number {
  if (b === 0) {
    return a;
  }

  return gcd(b, a % b);
}

File: solutions/csharp/rational-numbers/1/RationalNumbers.cs
public static class RealNumberExtension
{
    public static double Expreal(this int realNumber, RationalNumber r)  {
        var num = r.Numerator;
        var den = r.Denominator;

        return Root(Math.Pow(realNumber, num), den);
    }
    
    private static double Root(double p, double q) => Math.Pow(p, 1.0 / q);
}

public struct RationalNumber(int numerator, int denominator)
{
    public int Numerator { get; private set; } = numerator;
    public int Denominator { get; private set; } = denominator;

    public static RationalNumber operator +(RationalNumber r1, RationalNumber r2)
    {
        var a1 = r1.Numerator;
        var a2 = r2.Numerator;
        var b1 = r1.Denominator;
        var b2 = r2.Denominator;

        var num = (a1 * b2 + a2 * b1);
        var den = (b1 * b2);
        
        return new RationalNumber(num, den).Reduce();
    }

    public static RationalNumber operator -(RationalNumber r1, RationalNumber r2)
    {
        var a1 = r1.Numerator;
        var a2 = r2.Numerator;
        var b1 = r1.Denominator;
        var b2 = r2.Denominator;
        
        var num = (a1 * b2 - a2 * b1);
        var den = (b1 * b2);
        
        return new RationalNumber(num, den).Reduce();
    }

    public static RationalNumber operator *(RationalNumber r1, RationalNumber r2)
    {
        var a1 = r1.Numerator;
        var a2 = r2.Numerator;
        var b1 = r1.Denominator;
        var b2 = r2.Denominator;

        var num = a1 * a2;
        var den = b1 * b2;
        
        return new  RationalNumber(num, den).Reduce();
    }

    public static RationalNumber operator /(RationalNumber r1, RationalNumber r2)
    {
        var a1 = r1.Numerator;
        var a2 = r2.Numerator;
        var b1 = r1.Denominator;
        var b2 = r2.Denominator;

        var num = (a1 * b2);
        var den = (a2 * b1);
        
        return new RationalNumber(num, den).Reduce();
    }

    public RationalNumber Abs()
    {
        var result = new RationalNumber
        {
            Numerator = Math.Abs(Numerator),
            Denominator = Math.Abs(Denominator)
        };
        
        return result.Reduce();
    }

    public RationalNumber Reduce()
    {
        var num = Math.Abs(Numerator);
        var den = Math.Abs(Denominator);
        
        var gcd =  Gcd(num, den);
        

        var isPos = (Numerator >= 0 && Denominator >= 0) || (Numerator < 0 && Denominator < 0);
        
        Numerator = num / gcd * (isPos ? 1 : -1);
        Denominator = den / gcd;

        return this;
    }

    public RationalNumber Exprational(int power)
    {
        var num = Numerator;
        var den = Denominator;

        if (power > 0)
        {
            return new RationalNumber(
                (int)Math.Pow(num, power),
                (int)Math.Pow(den, power)
                ).Reduce();
        }
        
        return new RationalNumber(
            (int)Math.Pow(den, Math.Abs(power)),
            (int)Math.Pow(num, Math.Abs(power))
        ).Reduce();
    }
    
    private static int Gcd(int a, int b)
    {
        while (true)
        {
            if (b == 0) return a;
            var a1 = a;
            a = b;
            b = a1 % b;
        }
    }
    
}